drivers/dma/dma_mcux_smartdma.c

```diff
@@ -0,0 +1,251 @@
+
+	/* SMARTDMA does not have channels, so we only support channel 0. */
+	if (channel != 0) {
+		return -ENOTSUP;
+	}
+
+	data->callback = config->dma_callback;
+	data->user_data = config->user_data;
+
+	/* Reset smartDMA */
+	SMARTDMA_Reset();
+
+	/*
+	 * The dma_slot parameter is used to determine which SMARTDMA program
+	 * to run. First, convert the Zephyr define to a HAL enum.
+	 */
+	switch (config->dma_slot) {
+	case DMA_SMARTDMA_MIPI_RGB565_DMA:
+		prog_idx = kSMARTDMA_MIPI_RGB565_DMA;
+		break;
+	case DMA_SMARTDMA_MIPI_RGB888_DMA:
+		prog_idx = kSMARTDMA_MIPI_RGB888_DMA;
+		break;
+	case DMA_SMARTDMA_MIPI_RGB565_180:
+		prog_idx = kSMARTDMA_MIPI_RGB565_R180_DMA;
+		break;
+	case DMA_SMARTDMA_MIPI_RGB888_180:
+		prog_idx = kSMARTDMA_MIPI_RGB888_R180_DMA;
+		break;
+	case DMA_SMARTDMA_MIPI_RGB565_DMA_SWAP:
+		swap_pixels = true;
+		prog_idx = kSMARTDMA_MIPI_RGB565_DMA;
+		break;
+	case DMA_SMARTDMA_MIPI_RGB888_DMA_SWAP:
+		swap_pixels = true;
+		prog_idx = kSMARTDMA_MIPI_RGB888_DMA;
+		break;
+	case DMA_SMARTDMA_MIPI_RGB565_180_SWAP:
+		swap_pixels = true;
+		prog_idx = kSMARTDMA_MIPI_RGB565_R180_DMA;
+		break;
+	case DMA_SMARTDMA_MIPI_RGB888_180_SWAP:
+		swap_pixels = true;
+		prog_idx = kSMARTDMA_MIPI_RGB888_R180_DMA;
+		break;
+	default:
+		prog_idx = config->dma_slot;
+		break;
+	}
+
+	switch (prog_idx) {
+	case kSMARTDMA_MIPI_RGB565_DMA:
+	case kSMARTDMA_MIPI_RGB888_DMA:
+	case kSMARTDMA_MIPI_RGB565_R180_DMA:
+	case kSMARTDMA_MIPI_RGB888_R180_DMA:
+		smartdma_dsi_param_t param = {.disablePixelByteSwap = (swap_pixels == false)};
+
+		if (config->block_count != 1) {
+			return -ENOTSUP;
+		}
+		/* Setup SMARTDMA */
+		param.p_buffer = (uint8_t *)config->head_block->source_address;
+		param.buffersize = config->head_block->block_size;
+		param.smartdma_stack = data->smartdma_stack;
+		/* Save configuration to SMARTDMA */
+		dev_config->base->ARM2SMARTDMA = (uint32_t)(&param);
+		break;
+	default:
+		/* For the default case, we simply pass the entire DMA config
+		 * struct to the SMARTDMA. The user's application could then
+		 * parse this structure, or choose to write different
+		 * configuration data to the SMARTDMA in their application
+		 */
+		dev_config->base->ARM2SMARTDMA = ((uint32_t)config);
+		break;
+	}
+	/* Save program */
+	dev_config->base->BOOT = (uint32_t)dev_config->smartdma_progs[prog_idx];
+	LOG_DBG("Boot address set to 0x%X", dev_config->base->BOOT);
+	return 0;
+}
+
+static int dma_mcux_smartdma_start(const struct device *dev, uint32_t channel)
+{
+	const struct dma_mcux_smartdma_config *config = dev->config;
+
+#ifdef CONFIG_PM
+	/* Block PM transition until DMA completes */
+	pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+#endif
+	/* Kick off SMARTDMA */
+	config->base->CTRL = SMARTDMA_MAGIC | SMARTDMA_BOOT;
+	return 0;
+}
+
+
+static int dma_mcux_smartdma_stop(const struct device *dev, uint32_t channel)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(channel);
+	/* Stop DMA  */
+	SMARTDMA_Reset();
```

Comment:
is the only way to stop with a reset of the peripheral?

Reply:
Indeed